client/task_runner.go

```diff
@@ -330,15 +380,53 @@ func (r *TaskRunner) SaveState() error {
 		snap.HandleID = r.handle.ID()
 	}
 	r.handleLock.Unlock()
-	return persistState(r.stateFilePath(), &snap)
+
+	// If nothing has changed avoid the write
+	h := snap.Hash()
+	if bytes.Equal(h, r.persistedHash) {
+		r.persistLock.Unlock()
+		return nil
+	}
+
+	// Serialize the object
+	var buf bytes.Buffer
+	if err := codec.NewEncoder(&buf, structs.MsgpackHandle).Encode(&snap); err != nil {
+		return fmt.Errorf("failed to serialize snapshot: %v", err)
+	}
+	r.persistLock.Unlock()
```

Comment:
Put back into defer as there's a race (albeit _tiny_) between this line and calling `stateDB.Batch()` that could reorder SaveState calls.